Implement multibase for C++

theoreticalbts opened the pull request:
Closes #168.
Resolves #169.

sgerbino reviewed (changes requested):
- We need more extensive coverage via testing of new code
- We shouldn't remove the bitcoin copyright
- The tests should be implemened as a Boost Test and run in CI

theoreticalbts commented:
The copyright wasn't removed, it was moved lower in the file.

For more clarity about what copyright applies to what code, I'm going to update the branch and put the code from Bitcoin in a separate header file from our code.

theoreticalbts commented:
Updated this PR:

- Addressed copyright issue
- Added tests
- Fixed several bugs uncovered by the tests

Also added a Monte Carlo test which you can run forever by:

```
    ../programs/test_multibase/multibase_mc.py | programs/test_multibase/test_multibase
```

You can insert a filter, e.g. `-head 1000` to do 1000 cases, or `pv -l` to see a progress bar.  The MC test has passed over 5 million cases with no failures.

mvandeberg commented on programs/koinos-types/lang/koinos_codegen_cpp/rt/json.hpp:
```diff
@@ -210,30 +210,19 @@ inline void from_json( const json& j, vector< T >& v, uint32_t depth )
-   util::encode_base58( base58, v );
-   j = std::move( 'z' + base58 );
+   std::vector<char> dest;
+   util::encode_multibase( v.data(), v.size(), dest );
```
We need to make the default variable_blob encoding base64

mvandeberg commented on programs/koinos-types/lang/koinos_codegen_cpp/rt/json.hpp:
```diff
@@ -304,29 +293,20 @@ inline void from_json( const json& j, array< T, N >& v, uint32_t depth )
-   util::encode_base58( base58, v );
-   j = std::move( 'z' + base58 );
+   std::vector<char> dest;
+   util::encode_multibase( v.data(), v.size(), dest );
```
We need to make the default fixed_blob encoding base64

mvandeberg commented on programs/koinos-types/lang/koinos_codegen_cpp/rt/multihash_json.hpp:
```diff
@@ -16,23 +16,21 @@ inline void to_json( json& j, const multihash& v )
+   std::string serialized_bytes = ss.str();
+
+   std::vector<char> encoded;
+   util::encode_multibase( serialized_bytes.c_str(), serialized_bytes.size(), encoded );
```
I know the default encoding is base58, but should we make that explicit here?

mvandeberg commented on programs/koinos-types/lang/koinos_codegen_cpp/rt/util/multibase.hpp:
```diff
@@ -0,0 +1,109 @@
+   memcpy( dest.data()+1, temp.data(), temp.size() );
+}
+
+/*
```
Is this commented code needed anymore? If not, can we clean it up?

mvandeberg commented on tests/cpp/src/multibase_test.cpp:
```diff
@@ -0,0 +1,131 @@
+}
+
+
+BOOST_AUTO_TEST_CASE( multibase_test )
```
These are all positive test cases. We need some specific negative test cases.

Some examples:

Illegal characters in an encoding.
Incorrect lengths (base64 with/without padding)
